service/storage_service.cc

```diff
@@ -4720,12 +4721,19 @@ future<> storage_service::do_cluster_cleanup() {
 }
 
 future<sstring> storage_service::wait_for_topology_request_completion(utils::UUID id, bool require_entry) {
+    rtlogger.debug("Start waiting for topology request completion (request id {})", id);
     while (true) {
+        auto c = _topology_state_machine.reload_count;
         auto [done, error] = co_await  _sys_ks.local().get_topology_request_state(id, require_entry);
         if (done) {
+            rtlogger.debug("Request with id {} is completed with status: {}", id, error.empty() ? sstring("success") : error);
             co_return error;
         }
-        co_await _topology_state_machine.event.when();
+        if (c == _topology_state_machine.reload_count) {
+            // wait only if the state was not reloaded while we were preempted
+            rtlogger.debug("Waiting for a topology event while waiting for topology request completion (request id {})", id);
+            co_await _topology_state_machine.event.when();
```

Comment:
The logic is inverted compared to normal condition variable usage. The idiom is

```
while !done
    co_await event.when(done || c != _topology_state_machine.reload_count)
    // handle either done or c != _topology_state_machine.reload_coount
end while
```

This is extensible (for new events) and easier to understand for people who understand condition variables.

Reply:
How is this suppose to work here? `done` will not change by itself. There is an (async) action that needs to be done to change it. And the action should happen as a result of `event` signalling.

If a predicate is always required for a CV why do we have `when()` that does not take one?

Reply:
done and reload_count are supposed to be set from the outside loop.

when() without a predicate is similar to pthread_cond_wait(). There, the user is supposed to provide an inner loop around the predicate. The version that accepts a predicate provides the inner loop.

Reply:
> done and reload_count are supposed to be set from the outside loop.

Why? Here `done` is not set in outside the loop. It is a status that is checked as a result of event triggering and `event.when()` waits for the even. What exactly do you propose to do?

> 
> when() without a predicate is similar to pthread_cond_wait(). There, the user is supposed to provide an inner loop around the predicate. The version that accepts a predicate provides the inner loop.

So here the code provides the loop around the predicate.